### DISCUSSION TITLE
Feasibility to use whole Fprime (FSW and GDS) in real satellite operation

### DISCUSSION THREAD

**Discussion by kitty99099** [nasa/fprime#3024] [Q&A]
## What do I want to do?
Hi! I'm a satellite enthusiast. I started to learn Fprime since I want to group a college team to build a real CubeSat to be launched to LEO. Below is my preliminary blueprint.

![image](https://github.com/user-attachments/assets/4f1e5c23-ebb1-4b78-8e0b-62eb81b165dd)

## What can I do now?
I learned Fprime for almost half year. Now I think I have a certain kowledge of Fprime' basic running. Below is what I can do now. I used a RPI as the OBC and make sure the PC and the RPI to be connected to the same Wifi. Therefore I can use GDS interface supplied by Fprime to send command to and read data from the ADCS. I think I already have the ability to integrate more subsystems that a satellite needs, like EPS, TT&C, and Payload, as long as I have their user manual.

![image](https://github.com/user-attachments/assets/0949783b-8023-458c-a450-2e58c337a3be)

## What have I surveyed?
I haven't seen any CubeSat mission case keeping whole Fprime. In my observation, they might cut the fprime-gds which can be referred to [Ground Interface Architecture and Customization](https://nasa.github.io/fprime/UsersGuide/best/ground-interface.html) and connect to other ground station software like OpenMCT or Cosmos. However, I think that it would be too many works to do, so it may not be our first priority.

## What's my further blueprint now?
I referred to [Hub Pattern](https://nasa.github.io/fprime/UsersGuide/best/hub-pattern.html) and then designed the architecture below. I would like to build two deployments in which the CubeSat will be Node A and the ground station PC will be Node B, but I'm not sure its feasibility particularly I don't know if I can use RF as the transportation between two hubs. I would like to inquire your thoughts and advices. I will appreciate your any comments.

![image](https://github.com/user-attachments/assets/a9ff3346-0e49-4f79-ab3b-b9f111ec383b)

Below is the hub pattern architecture.

![image](https://github.com/user-attachments/assets/32f99f5f-b778-498d-836c-faacae6d246a)

I saw a [discussion](https://github.com/nasa/fprime/discussions/2639) which seems trying to solve the problem similar to mine, but there seems to be a little difference. Any comment or idea is appreciated!

I'm still learning OSI layer but not familiar with... Should it be known before the development?

## What have I known now? (Any advice or comment is appreciated)

1. I need to implement [Communication Adapter Interface](https://nasa.github.io/fprime/Design/communication-adapter-interface.html) to transmit framed data to and received framed data from my radio component.

**Comment by zimri-leisher** [comment]
RF is just a way to transmit arbitrary data, so yes, you could use two deployments and have the communicate via RF.
However, I do not see the need for a second deployment on the ground. The fprime-gds tool should be able to directly work with the output of the `TlmPacketizer` or `TlmChan` components, without any components in between. So usually, people send the output of the `TlmPacketizer/TlmChan` in the form of bytes straight to the radio -> antenna -> RF -> dish -> modem, which outputs bytes again, which are sent to fprime-gds.

**Reply by kitty99099** [reply]
Thank you @zimri-leisher for your insights. You gave me an easier architecture. I will figure it out how to develope without a second deployment on the ground. But do you know how to specifically modify the fprime native components to make it come true? I'll be grateful for your further advice.

**Comment by Joshua-Anderson** [comment]
> I haven't seen any CubeSat mission case keeping whole Fprime. In my observation, they might cut the fprime-gds which can be referred to [Ground Interface Architecture and Customization](https://nasa.github.io/fprime/UsersGuide/best/ground-interface.html) and connect to other ground station software like OpenMCT or Cosmos. However, I think that it would be too many works to do, so it may not be our first priority.

`fprime-gds` is designed with a lab environment in mind (i.e. talking to your cubesat in a lab or maybe some prototype hardware) so that you can hit the ground running and test your fprime application before you have a full ground software environment setup. It's missing very useful functionality like logging downlinked telemetry to a database and advanced telemetry plotting.

Generally projects swap out fprime-gds and integrate their chosen ground system like COSMOS. I'm not sure any project has used the fprime GDS for actual operations, but it's not uncommon to use some of the fprime-gds libraries as part of the integrate code needed to help handle the parsing and encoding of data heading to and from the spacecraft.

**Reply by kitty99099** [reply]
Thank you @Joshua-Anderson for your advice. I used to figure nothing out why fprime-gds is only suitable for lab test not until you gave some example about what useful functionality it may be missing. I'll reconsider about using other ground systems. Thank you!

**Comment by ReggieMarr** [comment]
Looks like a great outline, I've actually been working on a project which has some overlap here so I can add some comments.
First off I can join the chorus and say yes absolutely you can use RF here. As you mentioned here you can just implement your drivers as something compatible with the communication adapter interface and you should be able to easily swap out a TCP/UDP comms driver for your RF driver.

I would also say having a firm grasp of the OSI model is going to be very important here. That being said keep in mind the ["all models are wrong, some are useful"](https://en.wikiquote.org/wiki/George_E._P._Box) quote here. That is, it's a model that provides abstractions so you can get your head around the idealized separation of concerns but most (RE: all) protocol implementations will not necessarily map neatly onto the models layers.

WRT fprime-gds I will also chime in and say it's fantastic for prototype development and can get you going with minimal effort. That being said there are better, more robust options to use (such as the as of yet unmentioned [yamcs](https://yamcs.org/)).

As far as your topology goes I actually think the hub model is a good play here. I'm currently building off some existing work and developing data link support for CCSDS and one of my near term goals is to use that approach as part of the [proof of concept/prototype tracked here](https://github.com/ReggieMarr/space-vehicle-system/tree/using_ccsds_framers_deframers) so keep an eye out for when I make an MR.

**Reply by kitty99099** [reply]
Thank you @ReggieMarr for your comments. I will try yamcs. Do you know how to  swap TCP/UDP comms driver for my RF driver? Is  [this](https://github.com/nasa/fprime/tree/devel/Drv/Udp) the TCP/UDP comms driver? Should the RF driver be developed by myself? Thank you for your guidance. I'm not familiar with these, but I have strong energy to learn.

**Reply by ReggieMarr** [reply]
Take a look at the Ref [instances.fpp](https://github.com/nasa/fprime/blob/devel/Ref/Top/instances.fpp) and [topology.fpp](https://github.com/nasa/fprime/blob/devel/Ref/Top/topology.fpp), if you need a visualization I've linked the view that can be generated using ```fprime-util visualize``` in ```fprime/Ref/Top```.

Note that the ```comm``` instance is the end of the chain both for sending and receiving and the Ref is using a [TCPClient](https://github.com/nasa/fprime/blob/devel/Drv/TcpClient/TcpClient.fpp) instance (which would also be a fine starting point).

Uplink:
![Uplink(2)](https://github.com/user-attachments/assets/bc5c4b5b-77a8-42a0-ba58-7bbd08364e6a)

Downlink:
![Downlink(1)](https://github.com/user-attachments/assets/a671688f-5e46-478c-b3d6-a7951cd1e1d3)

You then need to create a driver/component which conforms to that interface; AKA provides implementations of those ports which are defined in the TCPClients fpp file (see this reference to the [ByteStreamDriverModel](https://github.com/nasa/fprime/blob/devel/Drv/ByteStreamDriverModel/docs/sdd.md) for more details).

Note that conformance to the interface just means that you have to have at least those ports/associated functions implemented.
You can always build more on top of it should your RF driver require it and the use of the ```include``` directive is useful here for that reason.

As an example you could take a look at the [loopback component I created](https://github.com/ReggieMarr/space-vehicle-system/blob/using_ccsds_framers_deframers/FlightComputer/CCSDSTester/CCSDSTester.fpp) which implements both the receiving and transmitting end of a mock comms driver which makes things easy for testing purposes. This can be seen in its visualization:

![NewDeploymentLink](https://github.com/user-attachments/assets/40d6dcc2-41a2-4682-84fa-836c4d3c4760)

Keep in mind that example is something I hacked together for a non-production proof of concept and has a number of known issues so don't take it as an example of a "good component" just an example of something which conforms to the interface.


As a next step I'd suggest you could create something similar, and then try and break things out with the hub model so you have 2 deployments talking to one another. 
Also to clarify my previous comment; The reason I suggested using the UDP interface for this is a) it's different than the usual TCPClient so will force you to think about what you're doing a bit more deeply and b) UDP is a connectionless interface which will be one step closer towards the nature of your RF link.

Hope that helps, and let me know if you get something working !
